fix handling page lead section (arch)

midik opened the pull request:
fixes #1139

midik commented:
Tested locally on archwiki with limited article list.
I'm not sure why codecov complains - there are no lines added (probably the reason is thet compiled JS got a few extra lines).

kelson42 commented:
@midik We should probably check if we have a JSON at all or an error? Because the process will otherwise just die later!

midik commented:
@kelson42 I don't think that does make a sense to check if we have just any `json` there - that sounds half measures as by me. However, validating json more closely there would make sense but that will degrade performance...

midik commented:
@kelson42 We don't rely to that force-local-parsoid flag anymore. Yes, we have a retry logic. However, we could get a number of consecutive errors and back-off strategy doesn't always get the proper result. I have rewritten the logic so that it picks a correct render method depends on the data. (More correctly, we decide to either render content or get pre-rendered one). So rendering logic now got separated into `render()` method of `Article` class where the action happens. I consider `Article` class as the entity where we could move the article-related logic while making the code clean and make the tests real unit. Also, while testing this, I moved one test that covers category-index article rendering to `jest` so that now it looks like unit test and now it takes 300 ms).
So this is not only fix for the particular bug but partial code reworking and simplification. However, this PR doesn't conflict with #1123 by files (even if they both are big) so I think I won't have a challenge to merge them both.

midik commented:
Travis is still failing because of S3 etag, not related to this changes. Moreover, we have plenty of failed builds last days, [even on master](https://travis-ci.com/github/openzim/mwoffliner/builds/168915033). Is that something that we working on in #1117 [here](https://github.com/openzim/mwoffliner/pull/1117/commits/b2d0466b4752779597613740a3548bebc0fc9907)?
@bakshiutkarsha please correct me if I'm wrong.

midik commented:
Here is two zims produced by this code for your convenience:
[Using local MCS and local Parsoid](https://kiwixoffline.slack.com/files/UVC5C964D/F014NPG6EF7/wikem_en_all_2020-06.zim.gz)
[Using REST API](https://kiwixoffline.slack.com/files/UVC5C964D/F014HDG5GAJ/wikipedia_sg_all_2020-06.zim)
(first one too big for github so I put them to slack)

midik commented:
Just found a bug with wpen-100
![image](https://user-images.githubusercontent.com/6795530/83562369-1cddf200-a522-11ea-9897-f96c836fda71.png)

midik commented:
> Just found a bug with wpen-100

Not related to these changes. Reproducible in [current master](https://github.com/openzim/mwoffliner/commit/ac7033402510ea24ca99fea5b5f74b21f5380b6e).

midik commented on src/Article/Article.ts:
```diff
@@ -0,0 +1,41 @@
+    ) as boolean;
+  }
+
+  // todo remove the flag (#1139)
```
Sorry, this is forgotten todo that's done. Removed.

midik commented on src/Article/Article.ts:
```diff
@@ -0,0 +1,41 @@
+  // todo remove the flag (#1139)
+  public async render(): Promise<RenderedArticle[]> {
+    this.details = await articleDetailXId.get(this.id);
+    if (this.isRendered) return await ArticleDesktopRenderer.render(this);
```
This flag came from assumption that we have two kind of data format from the remote API, in short: where we have html or don't. Based on this, we choose between "deskop" renderer (that ends up with simple extracting `html` property) and "mcs" renderer (that seems to be a renderer actually). This variable was intended to be an indicator of that.

> Only one page should be rendered with a DesktopRenderer... and this is the mainpage

I didn't know that. Source code hadn't match this statement at that point, and should be simplified even more then.

midik commented on src/Article/Article.ts:
```diff
@@ -0,0 +1,41 @@
+    this.json = json;
+    this.renderingOptions = renderingOptions;
+
+    this.isRendered = (
```
A value of this flag depends on the `read-only` props and marked as `read-only` itself. In Typescript classes, `read-only` means that the value set in constructor only (once) and never changed. A kind of class-level const. So there's no reasons to convert it to a method because a value will never change during the instance lifecycle.

kelson42 replied:
ok

midik commented on src/Article/ArticleDesktopRenderer.ts:
```diff
@@ -0,0 +1,52 @@
+      throw new Error(`Cannot render [${article.json}] into an article`);
+    }
+    if (article.json.visualeditor) {
+      return article.renderingOptions.isMainPage ? article.json.visualeditor.content : this.injectHeader(article.json.visualeditor.content, article.details);
```
you're right, it's odd. Was there due to historical reasons. Changed.
